opentelemetry/src/metrics/meter.rs

```diff
@@ -29,23 +29,18 @@ pub trait MeterProvider {
     /// let provider = global::meter_provider();
     ///
     /// // meter used in applications
-    /// let meter = provider.meter("my_app");
+    /// let meter = provider.meter("my_app".to_string());
     ///
     /// // meter used in libraries/crates that optionally includes version and schema url
     /// let meter = provider.versioned_meter(
-    ///     "my_library",
-    ///     Some(env!("CARGO_PKG_VERSION")),
-    ///     Some("https://opentelemetry.io/schema/1.0.0"),
+    ///     "my_library".to_string(),
+    ///     Some(env!("CARGO_PKG_VERSION").to_string()),
+    ///     Some("https://opentelemetry.io/schema/1.0.0".to_string()),
     ///     Some(vec![KeyValue::new("key", "value")]),
     /// );
     /// ```
-    fn meter(&self, name: impl Into<Cow<'static, str>>) -> Meter {
-        self.versioned_meter(
-            name,
-            None::<Cow<'static, str>>,
-            None::<Cow<'static, str>>,
-            None,
-        )
+    fn meter(&self, name: String) -> Meter {
```

Comment:
`fn meter(&self, name: &'static str)`

@TommyCpp @utpilla Can we restrict meter name to be static strings only? Are there scenarios which we prohibit with this? This should avoid the concern Zhongyang raised about reduced ergonomics, I think.

(I am generally motivated by `tracing` crates strategy of restricting things to 'static in many places. In this case, this is just meter name, so I think we can steal that strategy.)

Reply:
That should work! But insetad of `'static str` can we do `'a str`. IIUC we need to copy the string anyway so we just need to scope the lifetime of the string as long as the `meter` function. I expect most users to use static string so we will only impact users that doesn't using global module + doesn't use `'static str`

I was trying to think of any case someone won't have it as `'static str` . And the only viable cases (I can think of) is probably [fetching meter name from environment variables](https://doc.rust-lang.org/std/env/fn.var.html).

Reply:
Per my understanding having a lifetime parameter on inputs only doesn't impact object safety but let me know if the compiler starts yelling 🤣

Reply:
@TommyCpp Are you suggesting changing the API to this?

```diff
- fn meter(&self, name: impl Into<Cow<'static, str>>) -> Meter 
+ fn meter(&self, name: &str) -> Meter
```
Doing this, would still break users who could have been passing an owned `String` as the value for `name`. They would now have to update the call sites to use `String.as_str()`. Essentially, no matter how we remove generics (Into<T>) from argument type, we would end up breaking some or the other use case. Is that still okay with you?

Having said that, I'm okay with using `&str` as the argument type for meter `name`.

Reply:
& 'a str also suffer from same concern as this PR. I think all of us are agreeing for `&'static str`, so let us proceeed with that?
(The rare scenario where meter name comes from somewhere else, Box::leak can rescue. This should be acceptable as the scenario itself is rare and also, there is usually 1-2 meter in any app!).

Reply:
@TommyCpp Can you confirm if we can proceed with `&'static str`?
(A similar optimization can be ported to traces (and eventually logs) too)

Reply:
> and eventually logs

logs doesn't have these global components to remove to reduce the API surface.

Reply:
Logs - not today, but possible we need to expose it, as it maybe [promoted as a user facing API](https://github.com/open-telemetry/oteps/pull/265)

Reply:
> we would end up breaking some or the other use case

I am still leaning towards `'a str` as it's a good middleground IMO. I will skip on why `'a &str` or `'static &str` is better than `String` as it's clear to everyone. 

Comparing `'static &str` and `'a &str`. let's say you are a user that somehow are using `String` to fetch a meter. `s.to_str()` is a better change than [`s.leak()`](https://doc.rust-lang.org/nightly/alloc/string/struct.String.html#method.leak) as the `leak` function(as the name sugguest?) can cause memory leak 

Plus I think semantically there is no reason why the meter name needs to be `'static str` if it's used only to fetch the meter object

Reply:
> Comparing 'static &str and 'a &str. let's say you are a user that somehow are using String to fetch a meter. s.to_str() is a better change than [s.leak()](https://doc.rust-lang.org/nightly/alloc/string/struct.String.html#method.leak) as the leak function(as the name sugguest?) can cause memory leak

I don't fully follow the argument that it cause memory leak. Can you elaborate? Are there scenarios where someone keeps creating Meters repeatedly?

Here is my thinking:
1. Meter name is meant to indicate the scope of the instrumentation like package name, module name etc. They can always fit the `& 'static str`.
2. It is possible for someone to make meter name using dynamic strings, but I am not aware of any real reason why is that needed. There can be reasons like, "My meter name comes from env variable or an incoming http header etc.", but I don't think they are worth optimizing for. (TBH, I have never seen such needs ever).
3. Even if someone really needs 2, then they can use leak api, to keep the meter name alive for the rest of the application. The heap memory used for meter name will remain forever, even after dropping otel metrics completely. With static, the same amount of memory would have still been used, but just allocated in a different location. With &'a str, same amount of memory would be used (as we copy it), but this can be cleaned up with drop, but such create-cleanup-re-create-repeat is not a pattern we are promoting. 
4. If someone keeps creating new Meters for, say, each incoming request, then yes, they'll leak memory. But I don't think we need to worry about that. MeterProvider, Meter (and instruments) are meant to be created once, and reused. If users are repeatedly trying to create Meters, then that itself is an issue.

> Plus I think semantically there is no reason why the meter name needs to be 'static str if it's used only to fetch the meter object

I actually think 'static str does promote the idea that meter name is for scope, that is known ahead. Which is why I am proposing to make that case, irrespective of the use case shown in this PR. In other words, with or without this PR, I suggest to keep meter scope names 'static str only.

(I think we need to further exploring the idea of 'static str in other places like Attribute Keys, Instrument Names etc. but I don't have a strong argument to push for that, though `tracing` does that today.)

Reply:
Ah good point on the Meter name should be a compling value. Let's move forward with `'static str`. We can include pointer to the spec so users understand the constrain here

> I don't fully follow the argument that it cause memory leak. Can you elaborate

If we only consider the SDK we provided then there is no concern on memory leak. But there is no guarantee which SDK end users are using, depends on the underlying implementation it may retain the string more than needed. 

The tricky part is API can be used by library users. So we are basically forcing the library users to leak memory to an unknown implementation(as SDK is chosen by users). But if it's supposed to be a compiling value then there is no reason why library users shouldn't have a `'static str` already.

Reply:
Thanks. Lets proceed by making &'static for MeterName.